### DISCUSSION TITLE
Log-on rendering error, but still works

### DISCUSSION THREAD

**Discussion by GonzalesMatthew** [nss-evening-cohort-14/evening-client-side#330] [Capstones!] [answered]
Matthew Gonzales

No similar tickets.

When I log on to my app I get the popular error: 

![Screen Shot 2021-06-21 at 9 10 31 PM](https://user-images.githubusercontent.com/16343050/122851811-3bea2600-d2d5-11eb-899d-e66d8da876f1.png)

I inspected the Component Task.js where it says it's occurring and pinpointed it to my API calls with the help of this article which says that my asynchronous calls are setting values before my component mounts:

- https://www.robinwieruch.de/react-warning-cant-call-setstate-on-an-unmounted-component

It says in there to use logic to work around that. Something like: 

![Screen Shot 2021-06-21 at 9 14 38 PM](https://user-images.githubusercontent.com/16343050/122852117-b87d0480-d2d5-11eb-9588-e1af180ae7ff.png)

Basically it is creating a variable `_ismounted` that must be true before the the API call can set the response. This is where I get lost, for two reasons:

- It is using this React code we didn't cover, which I believe it comes from before React hooks were a thing? I googled what a constructor was and it appears to be the same thing as our favorite hook useState(). Also, componentDidMount() is essentially the same as our useEffect() hook, so now I'm doubly confused, because I feel like my code should be intrinsically performing the code in this article's example.
- Upon reading React documentation on useEffect() I split my API calls into separate concerns: before all my API GET calls were contained in one useEffect(). Since my calls are not essentially related I thought that might fix it, but I'm still getting the error.

https://reactjs.org/docs/hooks-effect.html

So, now I don't know what to try. Apparently it's not a major break, but can lead to performance issues, so I'd like to figure out how to correctly apply the fix from the first article, but using the latest React hooks.

Here is my src code:
https://github.com/GonzalesMatthew/CAPSTONE-matts-music-monitor/blob/main/src/components/Task.js

Here is my latest code block which separates concern:
```
  // get topic, subTopic, and instrument names for updateForm
  useEffect(() => {
    getTopic(rest.task.topicId).then(setTopic);
  }, []);
  useEffect(() => {
    getTopic(rest.task.subTopicId).then(setSubTopic);
  }, []);
  useEffect(() => {
    getInstrument(rest.task.instrumentId).then(setInstrument);
  }, []);
  // get Tascam info and Memo info for updateForm
  useEffect(() => {
    getTascam(rest.task.firebaseKey).then((resp) => {
      getMemo(resp[0].memoId1).then(setMemo1);
      getMemo(resp[0].memoId2).then(setMemo2);
      getMemo(resp[0].memoId3).then(setMemo3);
      setTascam(resp);
    });
  }, []);
```
Also I tried 
```
const [mount, setMount] = useState(false)
```
then setting mount to true inside of useEffect, and while my API calls no longer worked my error went away -- so I know it's for sure related to something about my API calls.

Thanks for reading all this!

**Comment by drteresavasquez** [comment, marked as answer]
@GonzalesMatthew will update with solution

**Reply by GonzalesMatthew** [reply]
Yes, it appears I was creating my hooks which run my app in App/index.js and then passing them through Routes.js and finally where I first, actually needed them in my UserProfile.js view. This caused my entire app to re-render whenever my hooks refreshed, starting from App/index.js down (basically the whole app!). The solution was to bring those hooks down into UserProfile.js where I first, actually needed them. This also enabled this component to be fully modularized and technically I could now use this component over and over error free.

Moral of the story: ask yourself if you actually need that hook there or if it can be brought lower into your app tree (else you will face the wrath of the popular 'unmounted component' error and incur performance issues).

A side problem was that I also needed to firebase.initialize in my root index.js and not my App/index.js.

Thanks, Dr. T!

**Comment by GonzalesMatthew** [comment]
Hello again,

May I get some emergency help on something I believe is related to this above fix? My DOM no longer re-renders when I make updates to my tasks. At first the DOM also didn't re-render when I added tasks, but I was able to fix that.

- I've spent about 2 and a half hours tonight trying to fix this bug and was planning on spending it rehearsing for my presentation instead - but I realized last night this bug and only had tonight to fix it.
- I'm thinking it has something to do with my either (a) my update logic in my function that handles form submissions, or (b) something technical in my update API call (which looks correct still), or (c) something technical on how my TaskWindow.js is handling the `tasks` prop (which also looks correct still). So I'm thinking it is something about my logic in my form submission but even that looks appropriate.
- Maybe we can review it in book club?
